pkg/kubelet/dockertools/manager.go

```diff
@@ -971,20 +977,57 @@ func (dm *DockerManager) checkVersionCompatibility() error {
 	return nil
 }
 
-func (dm *DockerManager) getDefaultSecurityOpt() ([]string, error) {
+func (dm *DockerManager) getSecurityOpt(pod *api.Pod, ctrName string) ([]string, error) {
 	version, err := dm.APIVersion()
 	if err != nil {
 		return nil, err
 	}
-	// seccomp is to be disabled on docker versions >= v1.10
+
+	// seccomp is only on docker versions >= v1.10
 	result, err := version.Compare(dockerV110APIVersion)
 	if err != nil {
 		return nil, err
 	}
-	if result >= 0 {
+	if result < 0 {
+		// return early for old versions
+		return nil, nil
+	}
+
+	profile, profileOK := pod.ObjectMeta.Annotations["security.alpha.kubernetes.io/seccomp/container/"+ctrName]
+	if !profileOK {
+		// try the pod profile
+		profile, profileOK = pod.ObjectMeta.Annotations["security.alpha.kubernetes.io/seccomp/pod"]
+		if !profileOK {
+			// return early the default
+			return defaultSecurityOpt, nil
+		}
+	}
+
+	if profile == "unconfined" {
+		// return early the default
 		return defaultSecurityOpt, nil
 	}
-	return nil, nil
+
+	if profile == "docker/default" {
+		// return nil so docker will load the default seccomp profile
+		return nil, nil
+	}
+
+	if !strings.HasPrefix(profile, "localhost") {
+		return nil, fmt.Errorf("unknown seccomp profile option: %s", profile)
+	}
+
+	file, err := ioutil.ReadFile(filepath.Join(dm.seccompProfileRoot, strings.TrimPrefix(profile, "localhost/")))
+	if err != nil {
+		return nil, err
+	}
+
+	b := bytes.NewBuffer(nil)
+	if err := json.Compact(b, file); err != nil {
+		return nil, err
+	}
+
+	return []string{fmt.Sprintf("seccomp=%s", b.Bytes())}, nil
```

Comment:
Yes it needs to be the file contents. The docker cli takes a file name,
reads its contents, and transmits the contents to the daemon.

On Wednesday, June 8, 2016, Dr. Stefan Schimanski <notifications@github.com>
wrote:

> In pkg/kubelet/dockertools/manager.go
> <https://github.com/kubernetes/kubernetes/pull/25324#discussion_r66230042>
> :
>
> > +
> > +	if !strings.HasPrefix(profile, "localhost") {
> > +		return nil, fmt.Errorf("unknown seccomp profile option: %s", profile)
> > +	}
> > +
> > +	file, err := ioutil.ReadFile(filepath.Join(dm.seccompProfileRoot, strings.TrimPrefix(profile, "localhost/")))
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +
> > +	b := bytes.NewBuffer(nil)
> > +	if err := json.Compact(b, file); err != nil {
> > +		return nil, err
> > +	}
> > +
> > +	return []string{fmt.Sprintf("seccomp=%s", b.Bytes())}, nil
>
> Should this really be the compacted file content or the filename itself?
> Compare https://docs.docker.com/engine/security/seccomp/
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/25324/files/aa8c72adaaab6cba51ff868516ac5b423a95f069#r66230042>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAABYoMSK4m547F8EFFS3otQ56-RjCKAks5qJpe1gaJpZM4IZwJt>
> .
>